### ISSUE TITLE
Passing context

### ISSUE THREAD

**Issue by axdg** [byteslice/ergo#3]
@husobee, it seems to me that using [net/context](https://godoc.org/golang.org/x/net/context) as the mechanism for passing values through the middleware chain is the most idiomatic way to go (net/context is bound to land in the standard lib soon). Echo and Gin both roll their own context's, an this makes interoperability of middleware functions impossible or least very messy. I assume you agree with this?

I've been reading some articles about how to go about implementing this ([Golang blog](https://blog.golang.org/context), [Joe Shaw blog](https://joeshaw.org/net-context-and-http-handler/)). I think that the default sig of an ergo middleware (and the sig used by, for instance, the default methodNotAllowed handler) should be:

```
func (context.Context, http.ResponseWriter) (err error)
```
A pointer to the request, and a pointer to the leaf node which was matched would be values stored in the context that was passed to the first handler. After that, the ability to set/get values from context would be dealt with in handlers and would not involve any API provided by ergo, users would have to use the functions available in the net/context package for this.

As well as this, I'd like to automatically wrap middleware that have common signatures (like echo does) such as: 
 - func(http.Handler) http.Handler
 - http.Handler
 - http.HandlerFunc
 - func(http.ResponseWriter, *http.Request)

Although obviously these middleware would not have access to the context.

I'm trying to flush out all of these ideas being I start implementing the middleware chaining, let me know what you think.

**Comment by husobee** [comment]
I think that signature would work well for ergo, and allow us to keep context through the middleware chain.

I think using net/context is definitely the correct choice for context. I also like how you can derive from a primitive context and cancel requests and implement timeouts/deadlines.

The concept of adding the leaf of the tree to the context is very sweet. 

I feel like we are ready to prototype the core middleware chaining ideas and initial routing.  How should we break up the work?

**Comment by axdg** [comment]
Yeh, it would be so useful to have the leaf node available to the handler, I'm not sure why none of the radix-router frameworks have done this.

One more thing in terms of design - Echo and Gin have helper methods for their contexts, like c.String(status, body) etc. I don't think that ergo should do this. I'd prefer that once the core is finished, we write a submodule that contains static helper function that can be called on the context if desired. 

Prototyping - So I think there should be an Ergo struct definition in an ergo.go file (just like echo), and the http.server and all of the logic required for wrapping / chaining middleware would be in there. Each ergo instance would allocate a new router and grab a pointer to it (as per Echo again). The router should reside in it's own file (multiplexor.go - more idiomatic).

I'd like to spend a few days putting together the router myself. I'm going to use Gin's insertion and lookup logic but I need to refactor the tree and node structure so that individual method handlers are stored in the leaf nodes. I'd like to figure out what the most efficient way to do this is (I suspect its by giving each node a handlers feild - [9]func(context.Context, http.ResponseWriter) array). 

Would you like to start on ergo.go? Let me know what you thought are.

**Comment by husobee** [comment]
Sounds good to me. Will get started on the main struct and handlers. Would like to recommend when we would like to merge a branch into master we do a PR, and the other person does a peer review and accepts the pull request.  Will that work?

**Comment by axdg** [comment]
@husobee of course, that was the point of starting the project with another author (and hopefully a few more in future) - I want everything being reviewed by another dev.

**Comment by husobee** [comment]
sweet.
